src/NServiceBus.Core/Routing/MessageDrivenSubscriptions/MessageDrivenSubscribeTerminator.cs

```diff
@@ -0,0 +1,97 @@
+    using System.Linq;
+    using System.Threading;
+    using NServiceBus.ConsistencyGuarantees;
+    using NServiceBus.DeliveryConstraints;
+    using NServiceBus.Logging;
+    using NServiceBus.Pipeline;
+    using NServiceBus.Routing;
+    using NServiceBus.Routing.MessageDrivenSubscriptions;
+    using NServiceBus.Transports;
+    using NServiceBus.Unicast.Queuing;
+    using NServiceBus.Unicast.Transport;
+
+    class MessageDrivenSubscribeTerminator : PipelineTerminator<SubscribeContext>
+    {
+        public MessageDrivenSubscribeTerminator(SubscriptionRouter subscriptionRouter, string replyToAddress, IDispatchMessages dispatcher)
+        {
+            this.subscriptionRouter = subscriptionRouter;
+            this.replyToAddress = replyToAddress;
+            this.dispatcher = dispatcher;
+        }
+
+        public override void Terminate(SubscribeContext context)
+        {
+            var eventType = context.EventType;
+
+            var publisherAddresses = subscriptionRouter.GetAddressesForEventType(eventType)
+                .ToList();
+
+            if (!publisherAddresses.Any())
+            {
+                State state;
+
+                if (!context.TryGet(out state) || string.IsNullOrEmpty(state.ExplicitPublisherAddress))
+                {
+                    throw new Exception(string.Format("No destination could be found for message type {0}. Check the <MessageEndpointMappings> section of the configuration of this endpoint for an entry either for this specific message type or for its assembly.", eventType));
+                }
+
+                publisherAddresses = new List<string> { state.ExplicitPublisherAddress };
+            }
+
+            foreach (var publisherAddress in publisherAddresses)
+            {
+                Logger.Debug("Subscribing to " + eventType.AssemblyQualifiedName + " at publisher queue " + publisherAddress);
+
+                var subscriptionMessage = ControlMessageFactory.Create(MessageIntentEnum.Subscribe);
+
+                subscriptionMessage.Headers[Headers.SubscriptionMessageType] = eventType.AssemblyQualifiedName;
+                subscriptionMessage.Headers[Headers.ReplyToAddress] = replyToAddress;
+
+                var address = publisherAddress;
+
+                ThreadPool.QueueUserWorkItem(state =>
+                    SendSubscribeMessageWithRetries(address, subscriptionMessage, eventType.AssemblyQualifiedName));
+            }
+        }
+
+        void SendSubscribeMessageWithRetries(string destination, OutgoingMessage subscriptionMessage, string messageType, int retriesCount = 0)
+        {
+            try
+            {
+                dispatcher.Dispatch(subscriptionMessage, new DispatchOptions(destination, new AtLeastOnce(), new List<DeliveryConstraint>()));
+            }
+            catch (QueueNotFoundException ex)
+            {
+                if (retriesCount < 10)
+                {
+                    Thread.Sleep(TimeSpan.FromSeconds(2));
```

Comment:
Since that was added in order to make it "just work" locally I think we can remove it as soon as we have the new routing sorted out. Since that wouldn't need this as all.